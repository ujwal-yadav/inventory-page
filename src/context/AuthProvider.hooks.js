import axios from 'axios';
import { useNavigate } from 'react-router-dom';

export const useAuthProvider = () => {
  const navigate = useNavigate();
  const token = localStorage.getItem('site');

  const login = async (data) => {
    try {
      const response = await axios.post('/login', data, {
        headers: {
          'Content-Type': 'application/json',
        },
      });

      const res = response.data;

      if (res.data) {
        localStorage.setItem('site', res.data.token);
        navigate('/');
      }
    } catch (err) {
      alert('Erorr while logging in!');
    }
  };

  const logOut = () => {
    localStorage.removeItem('site');
    navigate('/login');
  };

  return { token, login, logOut };
};
