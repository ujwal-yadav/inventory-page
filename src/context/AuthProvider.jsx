import { useContext, createContext } from 'react';
import { useAuthProvider } from './AuthProvider.hooks';
const AuthContext = createContext();

const AuthProvider = ({ children }) => {
  const { token, login, logOut } = useAuthProvider();

  return (
    <AuthContext.Provider value={{ token, login, logOut }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;

export const useAuth = () => {
  return useContext(AuthContext);
};
