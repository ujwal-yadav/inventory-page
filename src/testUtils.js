import { createTheme, ThemeProvider } from '@mui/material';
import { render } from '@testing-library/react';

export const renderWithTheme = (ui, { themeOptions = {} } = {}) => {
  const theme = createTheme(themeOptions);
  return render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>);
};
