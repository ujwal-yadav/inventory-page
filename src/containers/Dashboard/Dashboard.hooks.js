import { useEffect, useState } from 'react';
import { useAuth } from '../../context/AuthProvider';
import axios from 'axios';

export const useDashboard = () => {
  const [statsData, setStatsData] = useState([]);

  const { token } = useAuth();

  const fetchStatsData = async (url, title, color) => {
    const headers = {
      token,
    };

    try {
      const response = await axios.get(url, { headers });

      if (response.data) {
        setStatsData((prevStatsData) => [
          ...prevStatsData,
          {
            title,
            value: response.data.data?.length || 0,
            color,
          },
        ]);
      }
    } catch (err) {
      alert(`Error while fetching ${title} stats!`);
    }
  };

  useEffect(() => {
    fetchStatsData('/products', 'PRODUCTS', 'green');
    fetchStatsData('/sales-orders', 'ORDERS', 'brown');
    fetchStatsData('/customers', 'CUSTOMERS', 'blue');
    fetchStatsData('/deliveries', 'DELIVERIES', 'orange');
    fetchStatsData('/purchase-returns', 'RETURNS', 'hotpink');
  }, []);

  return { statsData };
};
