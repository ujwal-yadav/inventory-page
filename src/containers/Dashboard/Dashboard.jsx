import React from 'react';
import { Box, useMediaQuery, useTheme } from '@mui/material';
import InventoryTable from '../../components/InventoryTable/InventoryTable';
import StatsCard from '../../components/StatsCard/StatsCard';
import { useDashboard } from './Dashboard.hooks';

export const Dashboard = () => {
  const theme = useTheme();
  const isLargeView = useMediaQuery(theme.breakpoints.up('md'));

  const { statsData } = useDashboard();

  return (
    <Box sx={{ paddingX: isLargeView ? 8 : 2, paddingY: isLargeView ? 8 : 2 }}>
      <Box
        display={'flex'}
        flexWrap={'wrap'}
        gap={isLargeView ? 4 : 2}
        paddingBottom={isLargeView ? 3 : 2}
      >
        {statsData?.length > 0 &&
          statsData?.map((stat, index) => (
            <StatsCard
              key={stat.title}
              title={stat.title}
              value={stat.value}
              color={stat.color}
            />
          ))}
      </Box>
      <InventoryTable />
    </Box>
  );
};
