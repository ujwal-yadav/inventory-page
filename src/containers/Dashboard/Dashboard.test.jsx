import React from 'react';
import { renderHook, waitFor } from '@testing-library/react';
import { Dashboard } from './Dashboard';
import '@testing-library/jest-dom/extend-expect';
import { useDashboard } from './Dashboard.hooks';
import { renderWithTheme } from '../../testUtils';
import { useAuth } from '../../context/AuthProvider';
import axios from 'axios';

jest.mock('../../context/AuthProvider.jsx', () => ({
  useAuth: jest.fn(),
}));
jest.mock('axios');

beforeEach(() => {
  useAuth.mockReturnValue({ token: 'mock-token' });
});

describe('Dashboard', () => {
  it('should render correctly', () => {
    const { asFragment } = renderWithTheme(<Dashboard />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Dashboard- Hooks', () => {
  it('should fetch data and sets stats data correctly', async () => {
    const mockData = { data: { data: new Array(10) } };
    axios.get.mockResolvedValue(mockData);

    const { result } = renderHook(() => useDashboard());

    await waitFor(() => expect(result.current.statsData.length).toBe(5));

    expect(result.current.statsData).toEqual([
      { title: 'PRODUCTS', value: 10, color: 'green' },
      { title: 'ORDERS', value: 10, color: 'brown' },
      { title: 'CUSTOMERS', value: 10, color: 'blue' },
      { title: 'DELIVERIES', value: 10, color: 'orange' },
      { title: 'RETURNS', value: 10, color: 'hotpink' },
    ]);
  });

  it('should handle errors while fetching stats data', async () => {
    axios.get.mockImplementation((url) => {
      if (url === '/products') {
        return Promise.reject(new Error('Error'));
      }
      return Promise.resolve({ data: { data: [] } });
    });

    const alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {});

    const { result } = renderHook(() => useDashboard());

    await waitFor(() => expect(result.current.statsData.length).toBe(4));

    expect(alertMock).toHaveBeenCalledWith(
      'Error while fetching PRODUCTS stats!'
    );
    expect(result.current.statsData).toEqual([
      { title: 'ORDERS', value: 0, color: 'brown' },
      { title: 'CUSTOMERS', value: 0, color: 'blue' },
      { title: 'DELIVERIES', value: 0, color: 'orange' },
      { title: 'RETURNS', value: 0, color: 'hotpink' },
    ]);
  });
});
