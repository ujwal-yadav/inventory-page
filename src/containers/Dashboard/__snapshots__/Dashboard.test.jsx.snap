// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-thm7"
  >
    <div
      class="MuiBox-root css-142jpn8"
    />
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
    >
      Loading...
    </h6>
  </div>
</DocumentFragment>
`;
